### PR TITLE
CI: Run package/check on all packages

### PR DESCRIPTION
**Pull request by hauke**
This will check some basic properties on all package Makefiles after the full build. It checks for example if the PKG_HASH is correct and will fail if it is not correct. This should detect such problems before they get merged.

This will always be executed when we do a full package build, the check is pretty fast.

Previously located here: https://github.com/openwrt/openwrt/pull/12803

### CODE REVIEW

**Comment by hauke** [comment]
@Ansuel and @ynezz  How do I test this? I also would like to test this with a broken package Makefile.

**Comment by hauke** [comment]
I am now testing it here: https://github.com/hauke/openwrt/actions/runs/6343912946
https://github.com/hauke/openwrt/actions?query=branch%3Aci-run-package-check-test

**Comment by Ansuel** [comment]
@hauke now that i'm noticing it... I would put the check as a separate step and not after the build. What do you think? The check should be a quicker step

**Comment by hauke** [comment]
I will try this tomorrow

**Comment by ynezz** [comment]
> The check should be a quicker step

I'm using following elsewhere and thats basically all the steps which are needed to download and check the package integrity:
 
```yaml
    - make defconfig
    - make -j $(nproc) tools/tar/compile
    - make -j $(nproc) download check FIXUP=1

    - >
      git diff-index --exit-code HEAD || {
        ret=$?
        echo "Package integrity issues, please check packages-hash-issues.patch from artifacts"
        git diff | tee packages-hash-issues.patch
        exit $ret
      }
```

**Comment by hauke** [comment]
> @hauke now that i'm noticing it... I would put the check as a separate step and not after the build. What do you think? The check should be a quicker step

I placed it now into an own step. I am testing if it works:
https://github.com/hauke/openwrt/actions?query=branch%3Aci-run-package-check-test
https://github.com/hauke/openwrt/actions?query=branch%3Aci-run-package-check-test-pass

File: .github/workflows/reusable_build.yml [hunk 1]
```diff
@@ -578,3 +581,45 @@ jobs:
           name: ${{ inputs.target }}-${{ inputs.subtarget }}-external-toolchain
           path: openwrt/bin/targets/${{ inputs.target }}/${{ inputs.subtarget }}/${{ steps.get-toolchain-name.outputs.toolchain-name }}
           retention-days: 1
+
+  check:
+    name: Check packages for ${{ inputs.target }}/${{ inputs.subtarget }}
+    needs: build
```
Comment (ynezz) [inline .github/workflows/reusable_build.yml:587]: > and not after the build

I agree with that as well. The build itself is the most resource expensive process, so I would run it after this check, so it should be the opposite, build needs "Check packages".

**Comment by hauke** [comment]
I moved the check now to the beginning let's see if it works.
We have to select a configuration before t will only check selected packages. Maybe we can move the configuration set also into the setup step.

File: .github/workflows/reusable_build.yml [hunk 2]
```diff
@@ -137,9 +137,83 @@ jobs:
...
+
+      - name: Check packages
+        shell: su buildbot -c "sh -e {0}"
+        working-directory: openwrt
+        run: make download -j$(nproc) BUILD_LOG=1 || ret=$? .github/workflows/scripts/show_build_failures.sh
+
+      - name: Check packages
+        shell: su buildbot -c "sh -e {0}"
+        working-directory: openwrt
+        run: make package/check FIXUP=1 -j$(nproc) BUILD_LOG=1 || ret=$? .github/workflows/scripts/show_build_failures.sh
```
Comment (ynezz) [inline .github/workflows/reusable_build.yml:187]:
```suggestion
        run: make download check FIXUP=1 -j$(nproc) BUILD_LOG=1 || ret=$? .github/workflows/scripts/show_build_failures.sh
```

Comment (hauke) [reply .github/workflows/reusable_build.yml]: I combined this now.

**Comment by hauke** [comment]
I added the missing `make tools/tar/compile ` step. Otherwise it worked before. It found already a problem, see https://github.com/openwrt/openwrt/commit/954142f4774aa6c0dca0a44179b3b0d119782f38

**Comment by Ansuel** [comment]
@hauke i see this is only for check runs. With the quilt patch we can also add the refresh one?

**Comment by Ansuel** [comment]
Ok something we can think about later. Think I will merge this.

@ynezz Should we really block the build step if check fails?

**Comment by ynezz** [comment]
> Should we really block the build step if check fails?

Yes, would be nice to be aware, its just a CI for additional QA. Or do you've any argument against?

**Comment by Ansuel** [comment]
> > Should we really block the build step if check fails?
> 
> Yes, would be nice to be aware, its just a CI for additional QA. Or do you've any argument against?

nope nothing against, might be relevant when we will introduce patches check but I think it's better to me more conservative with runs

**Comment by Ansuel** [comment]
@hauke do we have variant like tools/check toolchain/check and packages/check? I think I will improve your version and by default run only the packages check (and make the toolchain one for the toolchain run and the tools in the tools workflow)

Also more broken packages came... Can you by chance take care of fixing them?

**Comment by hauke** [comment]
> @hauke do we have variant like tools/check toolchain/check and packages/check? I think I will improve your version and by default run only the packages check (and make the toolchain one for the toolchain run and the tools in the tools workflow)

Then we have to copy this code into other workflows, maybe make it shared?
Most hashes are for packages so it is good to have it there first. 
 
> Also more broken packages came... Can you by chance take care of fixing them?

I only saw one broken package and fixed it ( https://github.com/openwrt/openwrt/commit/954142f4774aa6c0dca0a44179b3b0d119782f38 ), are there more?
I haven't tested the feed.

**Comment by Ansuel** [comment]
> Then we have to copy this code into other workflows, maybe make it shared?
> Most hashes are for packages so it is good to have it there first.

In theory it would be placed in reusable_build and reusable_build_tools. So no idea if it's worth to generalize.

> I only saw one broken package and fixed it ( [openwrt/openwrt@954142f](https://github.com/openwrt/openwrt/commit/954142f4774aa6c0dca0a44179b3b0d119782f38) ), are there more?
> I haven't tested the feed.

yufut also had the hash broken. But took care of fixing it.

**Comment by hauke** [comment]
> > Then we have to copy this code into other workflows, maybe make it shared?
> > Most hashes are for packages so it is good to have it there first.
> 
> In theory it would be placed in reusable_build and reusable_build_tools. So no idea if it's worth to generalize.

I am fine if we detect such problems for the packages, we should not invest too much resources into this small topic. 

> 
> > I only saw one broken package and fixed it ( [openwrt/openwrt@954142f](https://github.com/openwrt/openwrt/commit/954142f4774aa6c0dca0a44179b3b0d119782f38) ), are there more?
> > I haven't tested the feed.
> 
> yufut also had the hash broken. But took care of fixing it.
thanks.

**Comment by Ansuel** [comment]
@hauke yep every change has already been done, if you are curious,  check the latest commits

**Pull request merged**